core/domain/user_services.py

```diff
@@ -2287,3 +2291,263 @@ def get_dashboard_stats(user_id):
         'num_ratings': num_ratings,
         'average_ratings': average_ratings
     }
+
+
+def _get_checkpoints_in_order(init_state_name, states):
+    """Returns the checkpoints of an exploration in sequential order by a
+    BFS traversal.
+
+    Args:
+        init_state_name: str. The name of the first state of the exploration.
+        states: dict(state). All states of the exploration.
+
+    Returns:
+        list(str). List of all checkpoints of the exploration in sequential
+        order.
+    """
+    queue = []
+    queue.append(init_state_name)
+    checkpoint_state_names = []
+    visited_state_names = []
+    while len(queue) > 0:
+        current_state_name = queue.pop()
+        if current_state_name not in visited_state_names:
+            visited_state_names.append(current_state_name)
+            current_state = states[current_state_name]
+            if (current_state.card_is_checkpoint and (
+                current_state_name not in checkpoint_state_names)):
+                checkpoint_state_names.append(current_state_name)
+            for answer_group in current_state.interaction.answer_groups:
+                queue.append(answer_group.outcome.dest)
+
+    return checkpoint_state_names
+
+
+def _get_most_distant_reached_checkpoint_in_current_exploration(
```

Comment:
most_distant or furthest would be better ?

Reply:
Including `furthest` in the method name might be confusing as it signifies we are using this to only update `furthest_reached_checkpoint`. This a generic function which we also using to update `most_recently_reached_checkpoint` (in current version) later on in `synchronize_learner_checkpoint_progress_with_current_exp_version` method.